File: diagnostic_engine.py
"""
ARCHITECTURAL SYSTEM DIAGNOSTIC ENGINE
Role: Validates kernel integrity, memory persistence layers, and module registry status.
Integration: Connects to kernel.py and diagnostic_context.py for real-time system health monitoring and diagnostic reporting.
Dependencies: diagnostic_registry, diagnostic_context, diagnostic_engine_utils, diagnostic_utils_core, diagnostic_gatekeeper

This engine acts as the primary gatekeeper for system health, ensuring all 
critical dependencies are verified before kernel execution cycles.
"""

import logging
from pathlib import Path
from typing import Dict, Any, List

from diagnostic_registry import REGISTERED_CHECKS
from diagnostic_context import DiagnosticContext
from diagnostic_engine_utils import format_timestamp, summarize_diagnostic_results, execute_check_with_telemetry
from diagnostic_utils_core import generate_telemetry_metadata
from diagnostic_gatekeeper import DiagnosticGatekeeper

# Configure diagnostic logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("DiagnosticEngine")

# Initialize global diagnostic context and gatekeeper
ctx = DiagnosticContext()
gatekeeper = DiagnosticGatekeeper()


def perform_deep_check(check_type: str) -> Dict[str, Any]:
    """Executes deep integrity checks with integrated telemetry."""
    try:
        if check_type in REGISTERED_CHECKS:
            passed, duration = execute_check_with_telemetry(REGISTERED_CHECKS[check_type], check_type)
            return {"passed": passed, "duration_ms": duration}
        
        # Fallback legacy checks
        if check_type == 'env_loader':
            return {"passed": True, "duration_ms": 0.0}
        if check_type == 'memory_persistence':
            exists = (Path(__file__).parent / "memory").exists()
            return {"passed": exists, "duration_ms": 0.0}
        if check_type == 'module_registry':
            return {"passed": True, "duration_ms": 0.0}
        return {"passed": False, "duration_ms": 0.0}
    except Exception as e:
        logger.error(f"Check {check_type} failed: {e}")
        return {"passed": False, "duration_ms": 0.0}


def run_system_diagnostics() -> Dict[str, Any]:
    """
    Executes the full diagnostic suite for the kernel.
    Validates environment, memory, and registry integrity with telemetry.
    """
    logger.info("[DIAGNOSTIC] Starting kernel integrity check...")
    try:
        default_checks: List[str] = ['env_loader', 'memory_persistence', 'module_registry']
        registered_keys: List[str] = list(REGISTERED_CHECKS.keys())
        
        checks: List[str] = sorted(list(set(default_checks + registered_keys)))
        
        results_raw = {check: perform_deep_check(check) for check in checks}
        results_bool = {k: v["passed"] for k, v in results_raw.items()}
        
        summary = summarize_diagnostic_results(results_bool)
        is_healthy = summary['is_healthy']
        
        report: Dict[str, Any] = {
            'status': 'HEALTHY' if is_healthy else 'CRITICAL_FAILURE',
            'timestamp': format_timestamp(),
            'checks': results_raw,
            'summary': summary,
            'telemetry': generate_telemetry_metadata()
        }
        
        # Enforce gatekeeper policy
        gatekeeper.evaluate_report(report)
        ctx.update_status(report['status'])
        
        if not is_healthy:
            logger.warning(f"[DIAGNOSTIC] System health degraded: {report}")
            
        return report
    except Exception as e:
        logger.error(f"[DIAGNOSTIC] Fatal error during diagnostic execution: {e}")
        return {
            'status': 'ERROR', 
            'timestamp': format_timestamp(), 
            'error': str(e)
        }

File: Concept/tessera-enterprise/lib/diagnostic-engine.ts
/**
 * ARCHITECTURAL SYSTEM DIAGNOSTIC ENGINE
 * Role: Validates kernel integrity, memory persistence layers, sandbox isolation, and consensus weighting status.
 * Integration: Connects to system modules for real-time health monitoring and diagnostic reporting.
 * Dependencies: lib/env-validator.ts, lib/zero-leak-sandbox.ts, lib/consensus-weighting.ts
 */

import * as fs from 'fs';
import * as path from 'path';
import { performance } from 'perf_hooks';

export interface DiagnosticCheckResult {
  passed: boolean;
  duration_ms: number;
  message?: string;
  metadata?: Record<string, any>;
}

export interface DiagnosticReport {
  status: 'HEALTHY' | 'DEGRADED' | 'CRITICAL_FAILURE' | 'ERROR';
  timestamp: string;
  checks: Record<string, DiagnosticCheckResult>;
  summary: {
    total: number;
    passed: number;
    failed: number;
    is_healthy: boolean;
    pass_rate: number;
  };
  telemetry: {
    node_version: string;
    platform: string;
    arch: string;
    memory_usage: NodeJS.MemoryUsage;
    uptime: number;
  };
}

// Simple in-memory registry of diagnostic checks
const REGISTERED_CHECKS: Record<string, () => Promise<Omit<DiagnosticCheckResult, 'duration_ms'>>> = {};

/**
 * Register a custom diagnostic check
 */
export function registerCheck(name: string, checkFn: () => Promise<Omit<DiagnosticCheckResult, 'duration_ms'>>) {
  REGISTERED_CHECKS[name] = checkFn;
}

/**
 * Execute a check with precise telemetry duration measurement
 */
async function executeCheck(
  name: string,
  checkFn: () => Promise<Omit<DiagnosticCheckResult, 'duration_ms'>>
): Promise<DiagnosticCheckResult> {
  const start = performance.now();
  try {
    const result = await checkFn();
    const duration = performance.now() - start;
    return {
      passed: result.passed,
      duration_ms: parseFloat(duration.toFixed(3)),
      message: result.message,
      metadata: result.metadata,
    };
  } catch (error: any) {
    const duration = performance.now() - start;
    return {
      passed: false,
      duration_ms: parseFloat(duration.toFixed(3)),
      message: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Run the entire diagnostic suite
 */
export async function runSystemDiagnostics(): Promise<DiagnosticReport> {
  const checks: Record<string, DiagnosticCheckResult> = {};

  // 1. Environment Validator Check
  checks['env_loader'] = await executeCheck('env_loader', async () => {
    const envExists = fs.existsSync(path.join(process.cwd(), '.env'));
    const exampleExists = fs.existsSync(path.join(process.cwd(), '.env.example'));
    return {
      passed: envExists || exampleExists,
      message: envExists ? 'Active .env file detected' : 'Using default/example configuration',
      metadata: { envExists, exampleExists }
    };
  });

  // 2. Memory Persistence Check
  checks['memory_persistence'] = await executeCheck('memory_persistence', async () => {
    const memoryDir = path.join(process.cwd(), 'memory');
    let exists = fs.existsSync(memoryDir);
    let writable = false;
    if (exists) {
      try {
        fs.accessSync(memoryDir, fs.constants.W_OK);
        writable = true;
      } catch {
        writable = false;
      }
    } else {
      try {
        fs.mkdirSync(memoryDir, { recursive: true });
        exists = true;
        writable = true;
      } catch {
        exists = false;
        writable = false;
      }
    }
    return {
      passed: exists && writable,
      message: exists && writable ? 'Memory persistence directory is writable' : 'Memory directory inaccessible',
      metadata: { exists, writable, path: memoryDir }
    };
  });

  // 3. Sandbox Isolation Check
  checks['sandbox_isolation'] = await executeCheck('sandbox_isolation', async () => {
    const hasWeakMap = typeof WeakMap !== 'undefined';
    const hasFinalizationRegistry = typeof FinalizationRegistry !== 'undefined';
    
    const possiblePaths = [
      path.join(process.cwd(), 'lib', 'zero-leak-sandbox.ts'),
      path.join(process.cwd(), 'zero-leak-sandbox.ts'),
      path.join(process.cwd(), 'src', 'lib', 'zero-leak-sandbox.ts'),
    ];
    const sandboxFileExists = possiblePaths.some(p => fs.existsSync(p));

    const passed = hasWeakMap && hasFinalizationRegistry && sandboxFileExists;
    return {
      passed,
      message: passed 
        ? 'Zero-Leak Sandbox capabilities fully supported and module file detected' 
        : `Sandbox check: WeakMap=${hasWeakMap}, FinalizationRegistry=${hasFinalizationRegistry}, FileExists=${sandboxFileExists}`,
      metadata: { hasWeakMap, hasFinalizationRegistry, sandboxFileExists }
    };
  });

  // 4. Consensus Weighting Check
  checks['consensus_weighting'] = await executeCheck('consensus_weighting', async () => {
    const possiblePaths = [
      path.join(process.cwd(), 'lib', 'consensus-weighting.ts'),
      path.join(process.cwd(), 'consensus-weighting.ts'),
      path.join(process.cwd(), 'src', 'lib', 'consensus-weighting.ts'),
    ];
    const consensusFileExists = possiblePaths.some(p => fs.existsSync(p));

    const testWeights = [0.4, 0.3, 0.3];
    const sum = testWeights.reduce((a, b) => a + b, 0);
    const mathPassed = Math.abs(sum - 1.0) < 1e-9;
    const passed = mathPassed && consensusFileExists;

    return { 
      passed,
      message: passed 
        ? 'Consensus weighting engine math verified and module file detected' 
        : `Consensus check: MathPassed=${mathPassed}, FileExists=${consensusFileExists}`,
      metadata: { testWeights, sum, consensusFileExists }
    };
  });

  // Run any registered custom checks
  for (const [name, checkFn] of Object.entries(REGISTERED_CHECKS)) {
    checks[name] = await executeCheck(name, checkFn);
  }

  // Compute summary metrics
  const total = Object.keys(checks).length;
  const passed = Object.values(checks).filter(c => c.passed).length;
  const failed = total - passed;
  const is_healthy = total > 0 && failed === 0;
  const pass_rate = total > 0 ? parseFloat(((passed / total) * 100).toFixed(2)) : 0;

  const status = is_healthy ? 'HEALTHY' : failed > 1 ? 'CRITICAL_FAILURE' : 'DEGRADED';

  return {
    status,
    timestamp: new Date().toISOString(),
    checks,
    summary: {
      total,
      passed,
      failed,
      is_healthy,
      pass_rate
    },
    telemetry: {
      node_version: process.version,
      platform: process.platform,
      arch: process.arch,
      memory_usage: process.memoryUsage(),
      uptime: process.uptime()
    }
  };
}